OpenCL function argument passing refactoring

dterrahe opened the pull request:
Inspired by this old comment
https://github.com/darktable-org/darktable/blob/086d7b2b82bf34061c7585f9df47b1433c11ec6c/src/common/locallaplaciancl.c#L194
this replaces this
```
  size_t sizes[] = { ROUNDUPDWD(p->width, devid), ROUNDUPDHT(p->height, devid), 1 };
  dt_opencl_set_kernel_arg(devid, kernel, 0, sizeof(cl_mem), (void *)&bl);
  dt_opencl_set_kernel_arg(devid, kernel, 1, sizeof(cl_mem), (void *)&bh);
  dt_opencl_set_kernel_arg(devid, kernel, 2, sizeof(int), (void *)&(width));
  dt_opencl_set_kernel_arg(devid, kernel, 3, sizeof(int), (void *)&(height));
  dt_opencl_set_kernel_arg(devid, kernel, 4, sizeof(float), (void *)&lpass_mult);
  err = dt_opencl_enqueue_kernel_2d(devid, kernel, sizes);

```
with this
```
  err = dt_opencl_enqueue_kernel_2d_args(devid, kernel, p->width, p->height,
    CLARG(bl), CLARG(bh), CLARG((width)), CLARG((height)), CLARG(lpass_mult));
```
This hopefully makes it simpler and less error prone to set up the link to opencl kernels.

The actual refactoring was done with a set of 7 regexps, with the first commit making sure that a `sizes` declared immediately before the first argument is not used later on, and so can be safely integrated in the "enqueue" call and deleted.

All arguments need to be wrapped in a CLANG or CLLOCAL macro, which takes care of the `sizeof` calculation (and adds some error checking, but of course it is still not aware of the actual opencl function parameters).

This should give identical results and my limited testing doesn't immediately show problems. A full regression test should be run before considering merging this, but I never manage to do this on my machine, so if somebody with a faster one all set up already would be willing to help, this would be _much_ appreciated!

The only actual change/bug fix (maybe; this is the 3rd commit) is in `nlmeans_core.c`; a reference to an array of two ints was passed, which now generates a warning that the `sizeof` is probably not what you want, whereas previously that was just ignored, as the compiler had no way of linking the two together. I can't quickly figure out what these functions are supposed to do and how to debug this, so it would be great if somebody more knowledgeable could have a look. It does seem that elsehwere `const int2` is just passed as the array (pointer) itself and not a reference to it.

dterrahe commented:
The regexps used for the actual conversion.
They were run over src/common,src/develop,src/iop search&replace.
The one that merges multiple args calls together needs to be run (at least) 6 times.

wrap "normal" argument:
dt_opencl_set_kernel_arg\((.*), (.*),[ ]*(.*),[ \n].*sizeof\(.*,[ \n]*.*&(.*)\);.*
dt_opencl_set_kernel_args($1, $2, $3, CLARG($4));

wrap local variable allocation:
dt_opencl_set_kernel_arg\((.*), (.*),[ ]*(.*),[ \n][ ]*(.*),[ \n][ ]*NULL\);.*
dt_opencl_set_kernel_args($1, $2, $3, CLLOCAL($4));

wrap "other" (array pointers usually):
dt_opencl_set_kernel_arg\((.*), (.*),[ ]*(.*),[ \n](.*),[ \n]*.*\(void \*\)(.*)\);.*
dt_opencl_set_kernel_args($1, $2, $3, CLWRAP($4, $5));

merge multiple arguments:
dt_opencl_set_kernel_args\((.*), (.*), (.*), (.*)\);\n[\n]*.*dt_opencl_set_kernel_args\(\1, \2, \d*, (.*)\);
dt_opencl_set_kernel_args($1, $2, $3, $4, $5);

swap kernel= and sizes= lines:
(.*ROUNDUPDWD.*ROUNDUPDHT.*\n)[\n]*(.*kernel =.*\n)
$2$1

bring sizes= line to the front:
(.*dt_opencl_set_kernel_args.*\n)(.*ROUNDUPDWD.*ROUNDUPDHT.*\n)(.*dt_opencl_enqueue_kernel_2d.*\n)
$2$1$3

move sizes= and args into enqueue:
([ ]*).*size_t (.*)\[.*=.*\{ ROUNDUPDWD\((.*), (.*)\), ROUNDUPDHT\((.*), \4\).*\};[\n]*
[ ]*dt_opencl_set_kernel_args\(\4, (.*), [ ]?0, (.*\n)[\n]*(.*)dt_opencl_enqueue_kernel_2d\(\4, \6, \2\);\n
$8dt_opencl_enqueue_kernel_2d_args($4, $6, $3, $5,\n$1  $7

jenshannoschwalm commented:
Whow!

TurboGit commented on src/common/opencl.h:
```diff
@@ -337,13 +337,33 @@ int dt_opencl_get_kernel_work_group_size(const int dev, const int kernel, size_t
                              const void *arg);
 
+/** wrap opencl arguments */
+#define CLWRAP(size, ptr) (const size_t)0xF111E8, (const size_t)size, (const void *)ptr
```
Ok, took me some time to understand this one. It is a marker to be able to parse the varargs in the implementation of `dt_opencl_enqueue_kernel_2d_args()`.

I would use a #define for this marker and add some comments to help future readers.

dterrahe replied:
> It is a marker to be able to parse the varargs

It is not really _needed_. It is a magic marker to prevent users from just putting variables directly when calling `dt_opencl_enqueue_kernel_2d_args` and forgetting to wrap them in CLARG or CLLOCAL. They should never use this marker by itself, because that would defeat the purpose of the extra check. If I add a #define for just the marker, that would increase the chance that somebody uses it incorrectly.

In theory, this marker could be used only in the debug version and be stripped out in the release version (so we'd put only 2 values on the stack for each argument instead of 3). But the "speedup" would of course be minimal. And I fear many developers don't use the debug version at all and this is meant to help _them_, so...

Yes, I could put all of this in a comment :-)

TurboGit commented:
> This should give identical results and my limited testing doesn't immediately show problems. A full regression test should be run before considering merging this, but I never manage to do this on my machine, so if somebody with a faster one all set up already would be willing to help, this would be _much_ appreciated!

I'll run the testsuite and will report.

TurboGit commented:
@dterrahe : The regression tests are all OK.

dterrahe commented:
Any comments on structure/naming conventions etc? Since most of this was auto generated, it should/could be relatively easy to make changes. Don't worry that I spend weeks doing this by hand and would be sensitive about criticism/suggestions.

TurboGit commented:
> Any comments on structure/naming conventions etc? Since most of this was auto generated, it should/could be relatively easy to make changes. Don't worry that I spend weeks doing this by hand and would be sensitive about criticism/suggestions.

Only one point about the generated code. If you could somehow split the lines it would be nice, the length is too much for me and a diff in a console will be hard.

dterrahe commented:
> If you could somehow split the lines it would be nice

That's the response I was expecting! :-)

I haven't figured out how to do this in a regexp yet. I'll see what I can find. 
If somebody could suggest some other tool please let me know!

TurboGit commented:
> That's the response I was expecting! :-)

A pleasure to see you happy :)

dterrahe commented:
While eyeballing the final result after splitting long lines, I noticed that in two cases a single parameter setting in an `else` was merged with the subsequent parameters (so that they would never be set in the `if` case.

There was also one case where in the original code in highlights.c one of the pararmeters in a call was actually set in a different kernel (so it didn't get merged with the others). That seemed "obviously wrong" to me, so I fixed it (in bd5c8a588ba9cece8eae8f2913053c390794a1e2) but it would still be good if this was reviewed by someone who actually understands opencl (same with db111b0c64dc5d16c495a02c50b668eb2f01a335).

Regression tests can't find everything...

jenshannoschwalm commented:
> That seemed "obviously wrong" to me,

In highlights: It was wrong :-) Thanks
The other one is also good.

TurboGit reviewed (changes requested):
While testing I got : 
`[local laplacian cl] failed: -51`

EDIT: with my change to display the error message we have:

```
12,860130 [local laplacian cl] couldn't enqueue kernel! CL_INVALID_ARG_SIZE
12,860174 [opencl_pixelpipe] [preview] could not run module `bilat' on gpu. falling back to cpu path
```

TurboGit commented on src/common/locallaplaciancl.c:
```diff
@@ -130,58 +130,40 @@ cl_int dt_local_laplacian_cl(
-    dt_opencl_set_kernel_arg(b->devid, b->global->kernel_gauss_reduce, 2, sizeof(int), &wd);
-    dt_opencl_set_kernel_arg(b->devid, b->global->kernel_gauss_reduce, 3, sizeof(int), &ht);
+      dt_opencl_set_kernel_args(b->devid, b->global->kernel_gauss_reduce, 1, CLARG(b->dev_padded[l]));
+    dt_opencl_set_kernel_args(b->devid, b->global->kernel_gauss_reduce, 2, CLARG(wd), CLARG(ht));
```
No correct as the arguments of dt_opencl_set_kernel_args() are `size` & `var`.

TurboGit replied:
You probably want here:

```
    dt_opencl_set_kernel_args(b->devid, b->global->kernel_gauss_reduce, 2, CLARG(wd));
    dt_opencl_set_kernel_args(b->devid, b->global->kernel_gauss_reduce, 3, CLARG(ht));
```

dterrahe replied:
That is (should be) the same thing; the multiple `args` version allows more than one argument (wrapped in CLARG etc) _starting_ at num.

TurboGit commented on src/common/locallaplaciancl.c:
```diff
@@ -130,58 +130,40 @@ cl_int dt_local_laplacian_cl(
 
   if(b->bwidth <= 1 || b->bheight <= 1) return err;
 
-  size_t sizes_pad[] = { ROUNDUPDWD(b->bwidth, b->devid), ROUNDUPDHT(b->bheight, b->devid), 1 };
```
Seems like the issue comes from both using sizes_pad. If I revert this chunk and the other one below I have no more errors.

TurboGit commented on src/common/locallaplaciancl.c:
```diff
@@ -130,58 +130,40 @@ cl_int dt_local_laplacian_cl(
   for(int k=0;k<num_gamma;k++)
   { // process images
     const float g = (k+.5f)/(float)num_gamma;
-    dt_opencl_set_kernel_arg(b->devid, b->global->kernel_process_curve, 0, sizeof(cl_mem), &b->dev_padded[0]);
```
This one should also be revert to avoid the issue.

That being said, I have no idea why at the moment.

dterrahe replied:
I have an untested theory that the problem is the last two arguments, b->bwidth and b->bheight. They are declared as `size_t`, but the `dt_opencl_set_kernel_arg` calls used to think they were `sizeof(int)` (=4) long. The new code ("correctly") sends their size as `sizeof(size_t)` =8, so maybe its not surprising that things fall over.

@jenshannoschwalm would there have been a good reason for the mismatch between variable sent and its claimed size? I suppose this could be fixed on the kernel side by correctly declaring the parameter as size_t (if that type is supported in cl) or in the definition of `dt_local_laplacian_cl_t`?

Maybe I'll be able to produce a dump of all variable sizes sent for all kernels in both before and after versions and if they line up that might be an easy way to check all in one go. Currently I'm trying to get pocl or oclgrind to work on my development machine (that doesn't have a GPU) so it is easier for me to test these things.

jenshannoschwalm replied:
I will have to look that up but from memory opencl size_t might be  not safe.

jenshannoschwalm replied:
Would be cl_long or cl_ulong

jenshannoschwalm replied:
I am just on mobile but indeed the size_t for pad looks wrong. The kernel seems to take int.

dterrahe replied:
size_t seems awfully big for one dimension of an image, but I guess the advantage is you can safely multiply without having to cast first. If I remember correctly @johnny-bit did a lot of work on that at some point, but enforcing consistency, where _all_ `width`s and `height`s would be size_t is a long way off. So then maybe having them all be ints and enforcing casts everywhere is the more consistent way? Or just deal with them case by case. Copy/pastes from a situation using size_t to one with ints will continue to introduce problems though.

At least I feel that this PR making this explicit, rather than just accepting mismatched `sizeof(type)`s and actual `type`s is "better"?

jenshannoschwalm replied:
Size_t is simply not required for image operations in CL code. Int as coordinates are fine.

TurboGit reviewed (commented):
I have also error reported with -d opencl on exposure module:

```
8,193075 [dt_opencl_check_tuning] use 3503MB (tunemem=ON, pinning=ON) on device `Quadro T1000' id=0
8,350365 [dt_opencl_enqueue_kernel_2d_with_local] kernel 24 on device 0: CL_INVALID_KERNEL_ARGS
8,351739 [opencl_blendop] couldn't enqueue kernel! CL_INVALID_KERNEL_ARGS
8,351744 [opencl_pixelpipe] [full] could not run module `exposure' on gpu. falling back to cpu path
8,553919 [dt_opencl_enqueue_kernel_2d_with_local] kernel 24 on device 0: CL_INVALID_KERNEL_ARGS
8,555488 [opencl_blendop] couldn't enqueue kernel! CL_INVALID_KERNEL_ARGS
```

dterrahe commented:
> error reported with -d opencl on exposure module

That seems to be in `blend.c`
I suspect `CLARG(offs)`, three times. That is not a correct translation of 
` dt_opencl_set_kernel_arg(devid, kernel_mask, 11, 2 * sizeof(int), (void *)&offs);`
Probably should be `CLARRAY(2, offs)`
Was the original "&" correct?

jenshannoschwalm commented:
&offs[0] ?

johnny-bit commented:
the size_t changes were mostly to make sure that any
multiplications/operations on things that should be both big and unsigned
stayed in realm of unsigned. That mostly started if i remember correctly
from maths error where big enough image could cause mem allocation request
to be netagive. That plus comparing signed and unsigned ints in C code can
lead to problems... If one makes sure to be cognizant of both, properly
handle casts & math issues and keep consitent then it's fine.

sob., 22 paź 2022 o 19:54 dterrahe ***@***.***> napisał(a):

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/common/locallaplaciancl.c
> <https://github.com/darktable-org/darktable/pull/12677#discussion_r1002546700>
> :
>
> >      err = dt_opencl_enqueue_kernel_2d(b->devid, b->global->kernel_gauss_reduce, sizes);
>      if(err != CL_SUCCESS) goto error;
>    }
>
>    for(int k=0;k<num_gamma;k++)
>    { // process images
>      const float g = (k+.5f)/(float)num_gamma;
> -    dt_opencl_set_kernel_arg(b->devid, b->global->kernel_process_curve, 0, sizeof(cl_mem), &b->dev_padded[0]);
>
> size_t seems awfully big for one dimension of an image, but I guess the
> advantage is you can safely multiply without having to cast first. If I
> remember correctly @johnny-bit <https://github.com/johnny-bit> did a lot
> of work on that at some point, but enforcing consistency, where *all*
> widths and heights would be size_t is a long way off. So then maybe
> having them all be ints and enforcing casts everywhere is the more
> consistent way? Or just deal with them case by case. Copy/pastes from a
> situation using size_t to one with ints will continue to introduce problems
> though.
>
> At least I feel that this PR making this explicit, rather than just
> accepting mismatched sizeof(type)s and actual types is "better"?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/darktable-org/darktable/pull/12677#discussion_r1002546700>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACRRKFPGOZZ67RX5TJ76JXLWEQS6DANCNFSM6AAAAAARKUHQTY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Pozdrawiam,
Hubert Kowalski

TurboGit commented:
> Probably should be `CLARRAY(2, offs)`

Sounds like it should yes !

TurboGit commented on src/common/bilateralcl.h:
```diff
@@ -35,7 +35,7 @@ typedef struct dt_bilateral_cl_t
   dt_bilateral_cl_global_t *global;
   int devid;
-  size_t size_x, size_y, size_z;
+  int size_x, size_y, size_z;
```
wondering if unsigned int won't be better here? Same question below.

dterrahe replied:
In theory, I guess yes. In practice, "all" width/height variables and struct members seem to be using `int`, so it would "look" weird and may lead to more surprising behavior than just having to keep one thing in your head. If it ever makes a difference anywhere...

Personally I'd rather be consistent than partially more "correct", but if you'd like me to change it, I won't object.

dterrahe commented:
I ran a before/after this refactor with all modules active and a few different blending options and got different argument sizes for these:

> splat
> blur_line
> blur_line_z
> slice
> pad_input
> process_curve
> colorreconstruction_splat
> colorreconstruction_blur_line
> colorreconstruction_slice

I looked into all of them and they all turned out to use size_t, from the definitions of shared structs. I changed those and checked if that would have an impact on any calculations using them. In a few cases I moved a multiplication with a sizeof in front, instead of after, it so that the type promotion would happen in time to give the same result. In many cases, the value in one of these size_ts gets simply assigned to int variables, so there's not much point in their added precision anyway.

Also made the `CLARRAY(2, offs)` change as discussed. I'm now no longer seeing any size differences.

BIG CAVEAT!!! My test clearly didn't pick up all kernels, as the mismatch requiring the CLARRAY changes didn't show up in the first place. I don't know if there is a runtime test to catch all opencl kernels? Some of them may be in deprecated modules. Maybe there's a style that uses all of them? (otherwise would be good to have for testing; but I wouldn't know how to create one except by painstakingly going over all the (combinations of) options in all the modules and blending and making sure all code is covered...)

TurboGit commented:
I've tested, here is what I've done to gain confidence that this does not break something.

I have run the regression testsuite without this PR and with this PR. I have then compared the logs. Note that the testsuite contains tests for most if not all modules even the deprecated ones.

```
$ diff -c logs/test-20221028-173323.log logs/test-20221028-175534.log 
*** logs/test-20221028-173323.log	2022-10-28 17:53:18.852979596 +0200
--- logs/test-20221028-175534.log	2022-10-28 18:15:26.841402568 +0200
***************
*** 528,534 ****
  
  Test 0016-lowpass-bilateral
        Image mire1.cr2
!       CPU & GPU version differ by 104786 pixels
        CPU vs. GPU report :
        ----------------------------------
        Max dE                   : 21.58795
--- 528,534 ----
  
  Test 0016-lowpass-bilateral
        Image mire1.cr2
!       CPU & GPU version differ by 104784 pixels
        CPU vs. GPU report :
        ----------------------------------
        Max dE                   : 21.58795
***************
*** 561,572 ****
  
  Test 0017-monochrome
        Image mire1.cr2
!       CPU & GPU version differ by 22809 pixels
        CPU vs. GPU report :
        ----------------------------------
        Max dE                   : 2.16921
        Avg dE                   : 0.00225
!       Std dE                   : 0.02646
        ----------------------------------
        Pixels below avg + 0 std : 99.18 %
        Pixels below avg + 1 std : 99.18 %
--- 561,572 ----
  
  Test 0017-monochrome
        Image mire1.cr2
!       CPU & GPU version differ by 22808 pixels
        CPU vs. GPU report :
        ----------------------------------
        Max dE                   : 2.16921
        Avg dE                   : 0.00225
!       Std dE                   : 0.02645
        ----------------------------------
        Pixels below avg + 0 std : 99.18 %
        Pixels below avg + 1 std : 99.18 %
***************
*** 627,637 ****
  
  Test 0019-color-mapping
        Image mire1.cr2
!       CPU & GPU version differ by 107356 pixels
        CPU vs. GPU report :
        ----------------------------------
        Max dE                   : 27.93468
!       Avg dE                   : 0.03419
        Std dE                   : 0.26758
        ----------------------------------
        Pixels below avg + 0 std : 96.16 %
--- 627,637 ----
  
  Test 0019-color-mapping
        Image mire1.cr2
!       CPU & GPU version differ by 107358 pixels
        CPU vs. GPU report :
        ----------------------------------
        Max dE                   : 27.93468
!       Avg dE                   : 0.03420
        Std dE                   : 0.26758
        ----------------------------------
        Pixels below avg + 0 std : 96.16 %
***************
*** 1716,1735 ****
  
  Test 0052-color-reconstruction
        Image mire1.cr2
!       CPU & GPU version differ by 888465 pixels
        CPU vs. GPU report :
        ----------------------------------
!       Max dE                   : 19.23852
!       Avg dE                   : 0.71821
!       Std dE                   : 1.40734
        ----------------------------------
!       Pixels below avg + 0 std : 75.01 %
!       Pixels below avg + 1 std : 86.67 %
!       Pixels below avg + 3 std : 97.33 %
        Pixels below avg + 6 std : 100.00 %
        Pixels below avg + 9 std : 100.00 %
        ----------------------------------
!       Pixels above tolerance   : 12.46 %
   
        Expected CPU vs. current CPU report :
        ----------------------------------
--- 1716,1735 ----
  
  Test 0052-color-reconstruction
        Image mire1.cr2
!       CPU & GPU version differ by 1.01173e+06 pixels
        CPU vs. GPU report :
        ----------------------------------
!       Max dE                   : 51.34364
!       Avg dE                   : 4.24255
!       Std dE                   : 7.87937
        ----------------------------------
!       Pixels below avg + 0 std : 75.54 %
!       Pixels below avg + 1 std : 82.56 %
!       Pixels below avg + 3 std : 98.66 %
        Pixels below avg + 6 std : 100.00 %
        Pixels below avg + 9 std : 100.00 %
        ----------------------------------
!       Pixels above tolerance   : 27.61 %
   
        Expected CPU vs. current CPU report :
        ----------------------------------
***************
*** 2508,2514 ****
  
  Test 0076-retouch-blur-fill
        Image mire1.cr2
!       CPU & GPU version differ by 143682 pixels
        CPU vs. GPU report :
        ----------------------------------
        Max dE                   : 16.46743
--- 2508,2514 ----
  
  Test 0076-retouch-blur-fill
        Image mire1.cr2
!       CPU & GPU version differ by 143681 pixels
        CPU vs. GPU report :
        ----------------------------------
        Max dE                   : 16.46743
***************
*** 2673,2684 ****
  
  Test 0081-mask-groups
        Image mire1.cr2
!       CPU & GPU version differ by 73776 pixels
        CPU vs. GPU report :
        ----------------------------------
        Max dE                   : 32.35596
        Avg dE                   : 0.03100
!       Std dE                   : 0.30957
        ----------------------------------
        Pixels below avg + 0 std : 97.36 %
        Pixels below avg + 1 std : 97.68 %
--- 2673,2684 ----
  
  Test 0081-mask-groups
        Image mire1.cr2
!       CPU & GPU version differ by 73751 pixels
        CPU vs. GPU report :
        ----------------------------------
        Max dE                   : 32.35596
        Avg dE                   : 0.03100
!       Std dE                   : 0.30956
        ----------------------------------
        Pixels below avg + 0 std : 97.36 %
        Pixels below avg + 1 std : 97.68 %
***************
*** 2739,2745 ****
  
  Test 0083-colorbalancergb
        Image mire1.cr2
!       CPU & GPU version differ by 27722 pixels
        CPU vs. GPU report :
        ----------------------------------
        Max dE                   : 1.33753
--- 2739,2745 ----
  
  Test 0083-colorbalancergb
        Image mire1.cr2
!       CPU & GPU version differ by 27723 pixels
        CPU vs. GPU report :
        ----------------------------------
        Max dE                   : 1.33753
```

Everything looks good except the test `0052-color-reconstruction`, the diff is huge so clearly a bug.

dterrahe commented:
Great! Thank you; I'll immediately look at that regression.

TurboGit commented on src/iop/colorreconstruction.c:
```diff
@@ -968,7 +952,7 @@ static cl_int dt_iop_colorreconstruct_bilateral_blur_cl(dt_iop_colorreconstruct_
 
   err = dt_opencl_enqueue_copy_buffer_to_buffer(b->devid, b->dev_grid, b->dev_grid_tmp, 0, 0,
-                                                b->size_x * b->size_y * b->size_z * 4 * sizeof(float));
+                                                sizeof(float) * b->size_x * b->size_y * b->size_z);
```
you've removed 4 here, I suspect this is the issue with test 0052.

TurboGit replied:
Ok, tested and I can confirm that it is the issue.

TurboGit reviewed (approved):
Thanks, working on my side and validated against the testsuite. We should be pretty safe, hopefully if it remains some breakages because of some holes in the testsuite we still have quite some filed testing.
